In [60]:
# Import libraries to use
import tensorflow as tf
import numpy as np
import keras as kr
import csv

# Split the data into training and testing

### Load data 

In [61]:
#data=np.genfromtxt('iris.csv',delimiter=',',dtype='str',header=None)
data=list(csv.reader(open('iris.csv')))[1:]
#list(data)
#print(data)

In [62]:
# Create inputs array with details and convert it from str to float
inputs  = np.array(data)[:,:4].astype(np.float)
# Create outputs array with species 
outputs = np.array(data)[:,4]
#list(inputs)
list(outputs)

['setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'setosa',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',
 'versicolor',


In [63]:
# return_inverse——https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.unique.html
# return the indices of the unique array ——each species gets a unique index in different position to represent their value 
outputs_vals, outputs_ints = np.unique(outputs, return_inverse=True)

# import this package to change type of index for next step that can match outputs data format 
from keras.utils import np_utils
outputs_cats = kr.utils.to_categorical(outputs_ints)
# test np.unique to return inverse value
print(outputs_cats)

[[ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1. 

In [64]:
# Split the input and output data sets into training and test subsets.
# reference to spilt data keras-iris——https://github.com/emerging-technologies/keras-iris/blob/master/iris_nn.py
# np.random.permutation——https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.random.permutation.html
# randomly  shuffle the elements of array


inds = np.random.permutation(len(inputs))

# make multiple sub-arrays to split data 75 by 75
train_inds, test_inds = np.array_split(inds, 2)

# match array of value to input and output type
inputs_train, outputs_train = inputs[train_inds], outputs_cats[train_inds]
inputs_test,  outputs_test  = inputs[test_inds],  outputs_cats[test_inds]
# test outputs data
print(outputs_test)

[[ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  0.  1.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]
 [ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]
 [ 0.  0.  1.]
 [ 1.  0. 

In [65]:
# Placeholders and variables. input has 4 attributes and output has 3 classes for species 
# x for inputs details with length of flowers and it can be fed data as same format
x=tf.placeholder(tf.float32,shape=[None,4])
# y for outputs details with species and it can be fed data as same format
y=tf.placeholder(tf.float32,shape=[None,3])
# Weight and bias
W=tf.Variable(tf.zeros([4,3]))
b=tf.Variable(tf.zeros([3]))

# Train model

In [66]:
# Model of softmax to predict results
prediction = tf.nn.softmax(tf.matmul(x, W) + b)

In [67]:
# cross_entropy function to calculate the cross entropy
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y, logits=prediction))

In [68]:
# Create optimiser to train
train_step = tf.train.AdamOptimizer(0.2).minimize(cross_entropy)

In [69]:
#initialising variables
init = tf.global_variables_initializer()

In [70]:
with tf.Session() as sess:
    sess.run(init)
# Number of interations
    for epoch in range(50):
        c=sess.run([train_step], feed_dict={x: inputs_train, y:outputs_train})

# Test model

### Calculating accuracy of our model 

In [71]:
# tf.equal to decide what value should be returned (True is 1,False is 0) 
correct_prediction = tf.equal(tf.argmax(prediction,1), tf.argmax(y,1))
# accuracy gets a average result of correct_prediction
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
# Test
with tf.Session() as sess:
    sess.run(init)
    # times for training all training set
    for epoch in range(75):    
        #train  
        sess.run(train_step,{x:inputs_train, y:outputs_train})                  
        #test
        acc = sess.run(accuracy,{x:inputs_test, y:outputs_test})     
        #print test value every 5 times begin from the first time
        if(epoch%5==0):
            print("Result " + str(epoch) + ",Testing Result " + str(acc) + ",Testing Error Rate " + str(1-acc))

Result 0,Testing Result 0.28,Testing Error Rate 0.719999998808
Result 5,Testing Result 0.64,Testing Error Rate 0.360000014305
Result 10,Testing Result 0.64,Testing Error Rate 0.360000014305
Result 15,Testing Result 0.64,Testing Error Rate 0.360000014305
Result 20,Testing Result 0.64,Testing Error Rate 0.360000014305
Result 25,Testing Result 0.733333,Testing Error Rate 0.266666650772
Result 30,Testing Result 0.96,Testing Error Rate 0.0400000214577
Result 35,Testing Result 0.773333,Testing Error Rate 0.226666688919
Result 40,Testing Result 0.8,Testing Error Rate 0.199999988079
Result 45,Testing Result 0.96,Testing Error Rate 0.0400000214577
Result 50,Testing Result 0.96,Testing Error Rate 0.0400000214577
Result 55,Testing Result 0.946667,Testing Error Rate 0.0533333420753
Result 60,Testing Result 0.946667,Testing Error Rate 0.0533333420753
Result 65,Testing Result 0.973333,Testing Error Rate 0.0266666412354
Result 70,Testing Result 0.96,Testing Error Rate 0.0400000214577


# The Accurency generates different results and accuracy rate keep at 96%